mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java

```diff
@@ -98,21 +101,36 @@ public CassandraMessageIdMapper(MailboxMapper mailboxMapper, AttachmentMapper at
                 .map(messageId -> imapUidDAO.retrieve((CassandraMessageId) messageId, Optional.empty())))
             .thenApply(stream -> stream.flatMap(Function.identity()))
             .thenApply(stream -> stream.collect(Guavate.toImmutableList()))
-            .thenCompose(composedMessageIds -> messageDAO.retrieveMessages(composedMessageIds, fetchType, Optional.empty())).join()
-            .filter(pair -> mailboxExists(pair.getLeft()))
+            .thenCompose(composedMessageIds -> messageDAO.retrieveMessages(composedMessageIds, fetchType, Optional.empty()))
+            .thenCompose(stream -> CompletableFutureUtil.allOf(
+                stream.map(pair -> mailboxExists(pair.getLeft())
+                    .thenApply(b -> backToPair(pair, b)))))
+            .join()
+            .flatMap(OptionalConverter::toStream)
             .map(loadAttachments(fetchType))
             .map(toMailboxMessages())
             .sorted(Comparator.comparing(MailboxMessage::getUid));
     }
 
-    private boolean mailboxExists(CassandraMessageDAO.MessageWithoutAttachment messageWithoutAttachment) {
-        try {
-            mailboxMapper.findMailboxById(messageWithoutAttachment.getMailboxId());
-            return true;
-        } catch (MailboxException e) {
-            LOGGER.info("Mailbox {} have been deleted but message {} is still attached to it.", messageWithoutAttachment.getMailboxId(), messageWithoutAttachment.getMessageId());
-            return false;
+    private Optional<Pair<CassandraMessageDAO.MessageWithoutAttachment, Stream<CassandraMessageDAO.MessageAttachmentRepresentation>>> backToPair(Pair<CassandraMessageDAO.MessageWithoutAttachment, Stream<CassandraMessageDAO.MessageAttachmentRepresentation>> pair, Boolean b) {
```

Comment:
Please read final diff.

Current code handle it like this : 

```
optional.thenApply(b -> Optional.of(pair).filter(any -> b)))))
```

Which is I belive better than what you propose.